staging/src/k8s.io/apiserver/pkg/storageversion/manager.go

```diff
@@ -112,22 +116,37 @@ func (s *defaultManager) UpdateStorageVersions(kubeAPIServerClientConfig *rest.C
 	sc := clientset.InternalV1alpha1().StorageVersions()
 
 	s.mu.RLock()
-	resources := []*ResourceInfo{}
+	resources := []ResourceInfo{}
 	for resource := range s.managedResourceInfos {
-		resources = append(resources, resource)
+		resources = append(resources, *resource)
 	}
 	s.mu.RUnlock()
 	hasFailure := false
-	for _, r := range resources {
+	// Sorting the list to make sure we have a consistent dedup result, and
+	// therefore avoid creating unnecessarily duplicated StorageVersion objects.
+	// For example, extensions.ingresses and networking.k8s.io.ingresses share
+	// the same underlying storage. Without sorting, in an HA cluster, one
+	// apiserver may dedup and update StorageVersion for extensions.ingresses,
```

Comment:
What happens if apiservers disagree about which groups are in what storage? e.g., I restart an apiserver with new flags that moves events?

(I actually think lots of things probably break but I want to know how it affects this system...)

Reply:
The storage migrator won't be able to tell from the storage versions, if you move events to a different storage. Say server A and B keep events in etcd 1, and now server C keep events in etcd 2. Their records look the same: 

```
apiVersion: internal.apiserver.k8s.io/v1alpha1
kind: StorageVersion
metadata:
  name: core.events
status:
  storageVersions:
  - apiServerID: A
    decodableVersions:
    - v1
    - events.k8s.io/v1beta1
    - events.k8s.io/v1
    encodingVersion: v1
  - apiServerID: B
    decodableVersions:
    - v1
    - events.k8s.io/v1beta1
    - events.k8s.io/v1
    encodingVersion: v1
  - apiServerID: C
    decodableVersions:
    - v1
    - events.k8s.io/v1beta1
    - events.k8s.io/v1
    encodingVersion: v1
```

so migration will be triggered even though it shouldn't. Now you have events stored in two places (assume the writes get load balanced) in the cluster. And when you read events from apisevers, different apiservers have different vision.

Reply:
Yeah, that's a disaster, but I think it already is. We can document that for now, you have to shut everything down and manually move things in etcd if you want to do this.

Reply:
ack. Will mention it in storage migrator doc.